File: problems/graphs/shortestPath.js
// const shortestPath = (edges, nodeA, nodeB) => {
  
//   let graph = buildGraph(edges)
//   let queue = [[nodeA, 0]]
//   let visited = new Set([nodeA])

//   while(queue.length) {

//     let [current, distance] = queue.shift()

//     if(current === nodeB) return distance 
//     for(let neighbor of graph[current]) {
//       if(!visited.has(neighbor)) {
//         visited.add(neighbor)
//         queue.push([neighbor, distance+1])
//       }

//     }
//   }
//   return -1
// };


const shortestPath = (edges, nodeA, nodeB) => {
  const graph = createGraph(edges)
  const visited = new Set()

  
  return explore(graph, nodeA, nodeB, visited)  


};


const explore = (graph, src, target, visited) => {
  
  let queue = [[src, 0]]
  while(queue.length) {
    let [current, count] = queue.shift()
    if(current===target) return count
    visited.add(current)
    for(let neighbor of graph[current]) {
      if(!visited.has(neighbor)) {
        queue.push([neighbor, count+1])
      }
    }
  }
  return -1
}

File: python_Structy/shortest_path.py
def shortest_path(edges, node_A, node_B):
  graph = _create_graph(edges)
  print(graph)

  result = _traverse(graph, node_A, node_B, set())
  if result == float('inf'): return -1 
  return result



def _traverse(graph, start, finish, visited):
  if start in visited: return float('inf')
  if start == finish: return 0

  visited.add(start)
  count = float('inf')

  for neighbor in graph[start]:
    count = min(_traverse(graph, neighbor, finish, set(visited)) + 1, count)

  return count
  

  

def _create_graph(edges):
  graph = {}

  for edge in edges: 
    x, y = edge 
    if x not in graph: graph[x] = []
    if y not in graph: graph[y] = []
    graph[x].append(y)
    graph[y].append(x)

  return graph